### PR TITLE
Flush free_list if it's stale

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Add an additional flag per CPU EBPF_EPOCH_PER_CPU_STALE.
This flag is accessed by ebpf_epoch_exit and by the flush timer.

This flag is set to on by the flush timer if the free_list is not empty.
This flag is set to off by ebpf_epoch_exit (if set).

If the flush timer sees this flag on and the free_list is not empty, then it will schedule an _ebpf_epoch_stale_worker DPC on the CPU (this DPC calls ebpf_epoch_enter/ebpf_epoch_exit).

Resolves: #813 

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

**Review by dthaler — APPROVED** [review]

File: libs/platform/ebpf_epoch.c [hunk 1]
```diff
@@ -445,7 +489,22 @@ _ebpf_epoch_get_release_epoch(_Out_ int64_t* release_epoch)
...
+        _ebpf_set_per_cpu_flag(&_ebpf_epoch_cpu_table[cpu_id], EBPF_EPOCH_PER_CPU_TIMER_ARMED, false);
+
+        if (!ebpf_list_is_empty(&_ebpf_epoch_cpu_table[cpu_id].free_list)) {
+            if (_ebpf_get_per_cpu_flag(&_ebpf_epoch_cpu_table[cpu_id], EBPF_EPOCH_PER_CPU_STALE)) {
+                // Schedule DPC
+                if (!_ebpf_epoch_cpu_table[cpu_id].stale_worker) {
+                    if (ebpf_allocate_non_preemptible_work_item(
+                            &_ebpf_epoch_cpu_table[cpu_id].stale_worker, cpu_id, _ebpf_epoch_stale_worker, NULL) ==
+                        EBPF_SUCCESS) {
+                        ebpf_queue_non_preemptible_work_item(_ebpf_epoch_cpu_table[cpu_id].stale_worker, NULL);
```
Comment (Alan-Jowett) [inline libs/platform/ebpf_epoch.c:501]: Bug:
Only queues it once after it's created. If it already exists, it doesn't queue it.

**Review by saxena-anurag — APPROVED** [review]

File: libs/platform/ebpf_epoch.c [hunk 2]
```diff
@@ -34,6 +34,8 @@
 // _ebpf_flush_timer:
 // Calls ebpf_flush and clears the _ebpf_flush_timer_set flag.
 //
+// _ebpf_epoch_stale_worker:
+// Called if a CPU has items to be freed, but hasn't run anything in EBPF_EPOCH_FLUSH_DELAY_IN_MICROSECONDS.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/platform/ebpf_epoch.c:38]: The value of 200us is picked as we don't want the free list to grow too large. A large free list will cause a delay on that CPU.

**Review by saxena-anurag — APPROVED** [review]

File: libs/platform/ebpf_epoch.c [hunk 3]
```diff
@@ -122,9 +121,48 @@ _ebpf_epoch_get_release_epoch(_Out_ int64_t* released_epoch);
...
+
+static bool
+_ebpf_get_per_cpu_flag(_In_ const ebpf_epoch_cpu_entry_t* cpu_entry, ebpf_epoch_per_cpu_flags_t flag)
+{
+    return (cpu_entry->flags & flag) != 0;
+}
+
+static void
+_ebpf_set_per_cpu_flag(_In_ ebpf_epoch_cpu_entry_t* cpu_entry, ebpf_epoch_per_cpu_flags_t flag, bool state)
+{
```
Comment (shankarseal) [inline libs/platform/ebpf_epoch.c:150]: consider using https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-interlockedor

Comment (Alan-Jowett) [reply libs/platform/ebpf_epoch.c:150]: Will fix this in another PR.

File: libs/platform/ebpf_epoch.c [hunk 4]
```diff
@@ -445,7 +490,23 @@ _ebpf_epoch_get_release_epoch(_Out_ int64_t* release_epoch)
...
 
         // Clear the flush timer flag.
-        _ebpf_epoch_cpu_table[cpu_id].flags.timer_armed = false;
+        _ebpf_set_per_cpu_flag(&_ebpf_epoch_cpu_table[cpu_id], EBPF_EPOCH_PER_CPU_TIMER_ARMED, false);
+
+        if (!ebpf_list_is_empty(&_ebpf_epoch_cpu_table[cpu_id].free_list)) {
+            if (_ebpf_get_per_cpu_flag(&_ebpf_epoch_cpu_table[cpu_id], EBPF_EPOCH_PER_CPU_STALE)) {
+                // Schedule DPC
+                if (!_ebpf_epoch_cpu_table[cpu_id].stale_worker) {
+                    ebpf_allocate_non_preemptible_work_item(
```
Comment (shankarseal) [inline libs/platform/ebpf_epoch.c:499]: why not do this at epoch_initiate? Are you avoiding to do this at initiate to save memory?

File: libs/platform/ebpf_epoch.c [hunk 5]
```diff
@@ -445,7 +490,23 @@ _ebpf_epoch_get_release_epoch(_Out_ int64_t* release_epoch)
...
+                        &_ebpf_epoch_cpu_table[cpu_id].stale_worker, cpu_id, _ebpf_epoch_stale_worker, NULL);
+                }
+                if (_ebpf_epoch_cpu_table[cpu_id].stale_worker) {
+                    ebpf_queue_non_preemptible_work_item(_ebpf_epoch_cpu_table[cpu_id].stale_worker, NULL);
+                }
+            } else {
+                _ebpf_set_per_cpu_flag(&_ebpf_epoch_cpu_table[cpu_id], EBPF_EPOCH_PER_CPU_STALE, true);
```
Comment (shankarseal) [inline libs/platform/ebpf_epoch.c:506]: Instead of waiting on another flush timer that may queue yet another DPC, why not set a timer here itself?

Comment (Alan-Jowett) [reply libs/platform/ebpf_epoch.c:506]: During the first cycle, the code marks the cpu entry as potentially stale. 
During the second cycle, if nothing has run on the cpu, schedule a DPC to free the work items.

File: libs/platform/ebpf_epoch.c (hunk 4 above)
Comment (Alan-Jowett) [reply libs/platform/ebpf_epoch.c:499]: If the code never runs on a CPU, we will never allocate a DPC for it.

**Pull request merged**